src/plugins/data_source_management/server/saved_objects/data_source_premission_client_wrapper.ts

```diff
@@ -0,0 +1,185 @@
+
+/**
+ * Determine whether the user has the permissions to create, delete, and update data source based on manageableBy/dataSourceAdmin.
+ * DataSourceAdmin user has all permissions.
+ * If manageableBy is all, any user has permissions.
+ * If manageableBy is none, any user has no permissions.
+ * If manageableBy is dashboard_admin, only OSD admin has permissions.
+ */
+export class DataSourcePermissionClientWrapper {
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const { isDashboardAdmin, isDataSourceAdmin } = getWorkspaceState(wrapperOptions.request) || {};
+
+    if (
+      isDataSourceAdmin ||
+      this.manageableBy === ManageableBy.All ||
+      (this.manageableBy === ManageableBy.DashboardAdmin && isDashboardAdmin !== false)
+    ) {
+      return wrapperOptions.client;
+    }
+
+    const createWithManageableBy = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      if (DATA_SOURCE_SAVED_OBJECT_TYPE === type) {
+        throw this.generatePermissionError();
+      }
+      return await wrapperOptions.client.create(type, attributes, options);
+    };
+
+    const bulkCreateWithManageableBy = async <T = unknown>(
+      objects: Array<SavedObjectsBulkCreateObject<T>>,
+      options?: SavedObjectsCreateOptions
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      const disallowedSavedObjects: Array<SavedObjectsBulkCreateObject<T>> = [];
+      const allowedSavedObjects: Array<SavedObjectsBulkCreateObject<T>> = [];
+
+      objects.forEach((item) => {
+        if (DATA_SOURCE_SAVED_OBJECT_TYPE === item.type) {
+          disallowedSavedObjects.push(item);
+          return;
+        }
+
+        allowedSavedObjects.push(item);
+        return;
+      });
+
+      const bulkCreateResult = await wrapperOptions.client.bulkCreate(allowedSavedObjects, options);
+
+      // Merge the data source saved objects and real client bulkCreate result.
+      return {
+        saved_objects: [
+          ...(bulkCreateResult?.saved_objects || []),
+          ...disallowedSavedObjects.map((item) => ({
+            ...item,
+            error: {
+              ...this.generatePermissionError().output.payload,
+              metadata: { isNotOverwritable: true },
+            },
+          })),
+        ],
+      } as SavedObjectsBulkResponse<T>;
+    };
+
+    const updateWithManageableBy = async <T = unknown>(
+      type: string,
+      id: string,
+      attributes: Partial<T>,
+      options: SavedObjectsUpdateOptions = {}
+    ): Promise<SavedObjectsUpdateResponse<T>> => {
+      if (DATA_SOURCE_SAVED_OBJECT_TYPE !== type) {
+        return await wrapperOptions.client.update(type, id, attributes, options);
+      }
+
+      // Although the manageableBy is none, the OSD admin can assign workspaces to the data source.
+      if (isDashboardAdmin !== false && options.workspaces) {
```

Comment:
```suggestion
      if (!!isDashboardAdmin && options.workspaces) {
```
`undefined !== false` => true

Reply:
If isDashboardAdmin is undefined, we consider the user is dashboard admin.